pinot-core/src/main/java/com/linkedin/pinot/core/minion/SegmentPurger.java

```diff
@@ -0,0 +1,241 @@
+  private int _numRecordsModified;
+
+  public SegmentPurger(@Nonnull File originalIndexDir, @Nonnull File workingDir, @Nullable RecordPurger recordPurger,
+      @Nullable RecordModifier recordModifier) {
+    Preconditions.checkArgument(recordPurger != null || recordModifier != null,
+        "At least one of record purger and modifier should be non-null");
+    _originalIndexDir = originalIndexDir;
+    _workingDir = workingDir;
+    _recordPurger = recordPurger;
+    _recordModifier = recordModifier;
+  }
+
+  public File purgeSegment() throws Exception {
+    SegmentMetadata segmentMetadata = new SegmentMetadataImpl(_originalIndexDir);
+    String tableName = segmentMetadata.getTableName();
+    String segmentName = segmentMetadata.getName();
+    LOGGER.info("Start purging table: {}, segment: {}", tableName, segmentName);
+
+    try (PurgeRecordReader purgeRecordReader = new PurgeRecordReader()) {
+      SegmentGeneratorConfig config = new SegmentGeneratorConfig(purgeRecordReader.getSchema());
+      config.setOutDir(_workingDir.getPath());
+      config.setTableName(tableName);
+      config.setSegmentName(segmentName);
+      // Keep index creation time the same as original segment because both segments use the same raw data.
+      // This way, for REFRESH case, when new segment gets pushed to controller, we can use index creation time to
+      // identify if the new pushed segment has newer data than the existing one.
+      config.setCreationTime(String.valueOf(segmentMetadata.getIndexCreationTime()));
+
+      StarTreeMetadata starTreeMetadata = segmentMetadata.getStarTreeMetadata();
+      if (starTreeMetadata != null) {
+        config.enableStarTreeIndex(StarTreeIndexSpec.fromStarTreeMetadata(starTreeMetadata));
+      }
+
+      SegmentIndexCreationDriverImpl driver = new SegmentIndexCreationDriverImpl();
+      driver.init(config, purgeRecordReader);
+      driver.build();
+
+      LOGGER.info("Finish purging table: {}, segment: {}, purged {} records, modified {} records", tableName,
+          segmentName, _numRecordsPurged, _numRecordsModified);
+      return new File(_workingDir, segmentName);
+    }
+  }
+
+  @VisibleForTesting
+  int getNumRecordsPurged() {
+    return _numRecordsPurged;
+  }
+
+  @VisibleForTesting
+  int getNumRecordsModified() {
+    return _numRecordsModified;
+  }
+
+  private class PurgeRecordReader implements RecordReader {
+    final PinotSegmentRecordReader _recordReader;
+
+    // Reusable generic row to store the next row to return
+    GenericRow _nextRow = new GenericRow();
+    // Flag to mark whether we need to fetch another row
+    boolean _nextRowReturned = true;
+    // Flag to mark whether all records have been iterated
+    boolean _finished = false;
+
+    PurgeRecordReader() throws Exception {
+      _recordReader = new PinotSegmentRecordReader(_originalIndexDir);
+    }
+
+    @Override
+    public boolean hasNext() {
+      if (_recordPurger == null) {
```

Comment:
The iterator logic seems too complex. Will the following work?
hasNext() {
     if (_recordPurger == null) {
        if (_recordReader.hasNext()) {
          _nextRow = _recordReader.next(_nextRow);
          return true;
        } else {
          return false;
        }
      } else {
        while (_recordReader.hasNext()) {
          _nextRow = _recordReader.next(_nextRow);
          if (_recordPurger.shouldPurge(_nextRow)) {
            _numRecordsPurged++;
            continue;
          }
          return true;
        }
        return false;
      }
}

next() {
    if (_recordModifier != null) {
        if (_recordModifier.modifyRecord(_nextRow)) {
          _numRecordsModified++;
        }
      }
      return _nextRow;
}

Reply:
For now it might work, but it will break the contract of iterator. If user call next() multiple times, it will return the same record without iterate to the next one. Also, calling next() should get a new object instead of reusing the same one.

Reply:
BTW, _finished flag is not mandatory, removed for simpler logic

Reply:
ok, without the finished flag, i am fine with this.